### PR TITLE
Add guidelines for observability in the components

### PR DESCRIPTION
**Pull request by jpkrohling**
As part of #2430, this PR adds guidelines for component developers on how to appropriately make their components observable.

﻿Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -232,6 +232,36 @@ the event happens.
 Make log message human readable and also include data that is needed for easier
 understanding of what happened and in what context.
 
+### Observability
+
+Out of the box, your users should be able to observe the state of your component. 
+The collector exposes a `/metrics` endpoint where your metrics will land.
```
Comment (bogdandrutu) [inline CONTRIBUTING.md:238]: Prometheus `/metrics` endpoint?

Comment (jpkrohling) [reply CONTRIBUTING.md:238]: OpenMetrics would be more accurate, I suppose.

**Comment by jpkrohling** [comment]
@bogdandrutu, would you please review/merge this one?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -232,6 +232,36 @@ the event happens.
 Make log message human readable and also include data that is needed for easier
 understanding of what happened and in what context.
 
+### Observability
+
+Out of the box, your users should be able to observe the state of your component. 
+The collector exposes an OpenMetrics endpoint at `/metrics` where your data will land.
```
Comment (tigrannajaryan) [inline CONTRIBUTING.md:238]: Is OpenMetrics the right term to use here or Prometheus? (I am not quite sure what the relation between the too is).

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -232,6 +232,36 @@ the event happens.
...
+- are you making HTTP calls? Measure the latency and expose a histogram. Consider
+providing a counter with the number of calls that your processor made.
+- are you making sampling decisions? Provide a counter with the outcomes.
+
+At this moment, the collector is still using the OpenCensus Go client for the 
+instrumentation, but that shouldn't stop you from making your component observable.
+As a good practice, have a `metrics.go` with your `stats` and `views`, registering
+the views in `./service/telemetry.go`.
+For the `opentelemetry-collector-contrib` repository, register your view in the
+factory until a more appropriate solution is in place.
```
Comment (tigrannajaryan) [inline CONTRIBUTING.md:257]: Is this premature to recommend? We have not yet decided how we want to handle this, so encouraging adding more custom metrics is going to make our lives more difficult if we end up deciding to do something that is different from OpenCensus Go metrics.

File: CONTRIBUTING.md (hunk 2 above)
Comment (jpkrohling) [reply CONTRIBUTING.md:238]: Not a metrics expert, but my understanding is that OpenMetrics is the standard that defines the format in which those metrics are exposed.

File: CONTRIBUTING.md (hunk 3 above)
Comment (jpkrohling) [reply CONTRIBUTING.md:257]: Yes, based on the discussion from last Wednesday, I think we are still defining what's the recommendation. I wrote this before the discussion, under the assumption that this was indeed the recommendation.

Comment (tigrannajaryan) [reply CONTRIBUTING.md:257]: So, maybe leave out the custom metrics for now and keep the rest?

Comment (jpkrohling) [reply CONTRIBUTING.md]: Done

**Comment by jpkrohling** [comment]
@bogdandrutu, @tigrannajaryan, kindly re-review this one, please.

**Comment by jpkrohling** [comment]
@tigrannajaryan, do you think this should be closed for now? I think the current state of this PR might be useful, but not complete.

**Comment by tigrannajaryan** [comment]
I think the first two paragraphs are fine. If you want to keep just that part then we can merge it.

**Comment by jpkrohling** [comment]
PR updated.

**Comment by jpkrohling** [comment]
ping @tigrannajaryan

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -246,6 +246,15 @@ the event happens.
 Make log message human readable and also include data that is needed for easier
 understanding of what happened and in what context.
 
+### Observability
+
+Out of the box, your users should be able to observe the state of your component. 
+The collector exposes an OpenMetrics endpoint at `/metrics` where your data will land.
```
Comment (tigrannajaryan) [inline CONTRIBUTING.md:252]:
```suggestion
The collector exposes an OpenMetrics endpoint at `http://localhost:8888/metrics` where your data will land.
```

**Review by tigrannajaryan — APPROVED** [review]

File: CONTRIBUTING.md (hunk 4 above)
Comment (jpkrohling) [reply CONTRIBUTING.md]: Done

**Pull request merged**